lib/spack/spack/spec.py

```diff
@@ -1780,18 +1781,22 @@ def read_yaml_dep_specs(dependency_dict):
         for dep_name, elt in dependency_dict.items():
             if isinstance(elt, six.string_types):
                 # original format, elt is just the dependency hash.
-                dag_hash, deptypes = elt, ['build', 'link']
+                dag_hash, deptypes, virtuals = elt, ['build', 'link'], []
             elif isinstance(elt, tuple):
                 # original deptypes format: (used tuples, not future-proof)
                 dag_hash, deptypes = elt
+                virtuals = []
             elif isinstance(elt, dict):
                 # new format: elements of dependency spec are keyed.
                 dag_hash, deptypes = elt['hash'], elt['type']
+                # FIXME: Document why we are using "get"
+                # FIXME: instead of subscripts (ci.py)
+                virtuals = elt.get('virtuals', [])
```

Comment:
@scottwittenburg I had to use `elt.get` temporarily because of a couple of tests in `ci.py` that hardcode a yaml representation of a spec as a literal byte string:

https://github.com/spack/spack/blob/e9e2d1d3cab1352e0b2dbaafb31aaaff499e6079/lib/spack/spack/test/ci.py#L87-L97

that is decoded in:

https://github.com/spack/spack/blob/e9e2d1d3cab1352e0b2dbaafb31aaaff499e6079/lib/spack/spack/ci.py#L817-L818

Since I added a field to the spec dictionary, this representation was failing when extracting the "virtuals" attribute on edges and right now I'm just providing a "default". I can update the encoded strings wherever they appear in the tests, etc. but is there some reason we are using them as arguments to some functions in `spack.ci`?

Reply:
In some cases the pipeline generation step needs to provide concrete spec to rebuild job (we can't tolerate the spec being concretized again at the time of the rebuild).  So we serialize the concrete spec during generation, and we pass it as a string in the generated yaml file.  Let me know if that answers your question or not.

Reply:
@alalazo: I don't think I'm understanding why adding an edge attribute here breaks the logic in `ci.py`.  Are the tests assuming the old format and getting confused by `provides:`?  Do they need to know anything about the contents of the YAML?

Reply:
> Are the tests assuming the old format and getting confused by provides:?

Yes. The issue is:
```console
>>> print(str(zlib.decompress(base64.b64decode(root_spec)).decode('utf-8')))
spec:
- bzip2:
    version: 1.0.8
    arch:
      platform: linux
      platform_os: centos7
      target: x86_64
    compiler:
      name: gcc
      version: 4.8.5
    namespace: builtin
    parameters:
      shared: true
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    dependencies:
      diffutils:
        hash: 7g5yr5g2pxmgo2iqx7faar5zn4nufx7m
        type:
        - build
    hash: hhbrbrhbf5tkbqcg57apuqeg2b7nhich
    build_hash: mwqnizponf2zgyfztp7l2z7xorbmqjv6
- diffutils:
    version: '3.7'
    arch:
      platform: linux
      platform_os: centos7
      target: x86_64
    compiler:
      name: gcc
      version: 4.8.5
    namespace: builtin
    parameters:
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    dependencies:
      libiconv:
        hash: 66scwnborhboy6vd554qx3p6ihe4c7y3
        type:
        - build
        - link
    hash: 7g5yr5g2pxmgo2iqx7faar5zn4nufx7m
    build_hash: 7g5yr5g2pxmgo2iqx7faar5zn4nufx7m
- libiconv:
    version: '1.16'
    arch:
      platform: linux
      platform_os: centos7
      target: x86_64
    compiler:
      name: gcc
      version: 4.8.5
    namespace: builtin
    parameters:
      cflags: []
      cppflags: []
      cxxflags: []
      fflags: []
      ldflags: []
      ldlibs: []
    hash: 66scwnborhboy6vd554qx3p6ihe4c7y3
    build_hash: 66scwnborhboy6vd554qx3p6ihe4c7y3
```
where `root_spec` is the byte blob posted above.

Reply:
@tgamblin @scottwittenburg I can regenerate the string of bytes and hard-code another blob here with the new field. Or better, I can code a fixture that generates the blob with the properties needed for the tests. Since I don't know this part of the code very well, I was just curious of the API for:

https://github.com/spack/spack/blob/e9e2d1d3cab1352e0b2dbaafb31aaaff499e6079/lib/spack/spack/ci.py#L795

In this function `root_spec` needs to be either a spec string or a blob of bytes depending on the value of `compiler_action`.

Reply:
I would just modify the test so that it round-trips from whatever the current spec representation looks like, and checks that the output is equal to the input.  If we regenerate the blob, this will remain brittle in the same way.

Reply:
@scottwittenburg @tgamblin Minimal fix in 0af013f. I think we can do some refactor to the code in the tests, but I would save it for a following PR, since this one is already huge!

Reply:
Thanks @alalazo that looks great 😃